### PR TITLE
Modernize to c++14 for gcc/clang like compilers

### PR DESCRIPTION
**Pull request by marscher**
So far, we have used c++14 standard only for the Windows platform, but Unix still used c++11.
This PR modernizes some constructs already introduced in c++11, but really not used so far. I've used clang-tidy to provide most of the fixes.

- Use "explicit" keyword for constructors to avoid implicitly casting something unrelated.
- Virtual constructors are now marked "override". Virtual and override together makes no sense (according to clang-tidy).
- prefer "using type_alias = type_x" semantics over "typedef"
- Use range operator to simplify statements
- Use "auto" where applicable
- use "nullptr" over macro NULL
- use variable initializers (avoids code duplication)
- use default ctors/dtors to shrink code and allow for further compiler optimization
- use std::move semantics (string moves apparently)
- mark unimplemented private special functions with "delete" and made them public
- fixed: uninitialized member vars for some classes (JPContext etc.)

In addition I cleaned up JPype.h to only contain STL related stuff, if it is mandatory to declare types used within headers. The implementation now includes what is really needed. This should speed up compilation a little bit.

Another important change is that JPypeException now derives from std::runtime_error, which adds some additional safety. The copy constructor of JPypeException was not marked "noexcept", so it could have thrown exceptions during the initialization (leading to a terminate call). As I didn't get the logic why the (string) message gets copied into a stringstream without any other additions just to make a string and then a cstring again, I just got rid of this behaviour.
By using a std exception type, we are safe that we use the proper standard definitions for exception safety. If this is undesired, I can also revert these changes.

### CODE REVIEW

**Comment by pelson** [comment]
Looks really extensive! Since there is such significant churn as a result, does it make sense to apply a formatter (e.g. ``clang-format``) to help with code consistency? (I don't propose it in this PR, but it could be a reasonable follow-on PR).

**Comment by Thrameos** [comment]
I would ask that we do not run a formatter.  It would kill epypj branch which is a huge work efforr that just needed people to help write tests.

**Comment by Thrameos** [comment]
@marscher thanks for tackling this one.  Unfortunately my knowledge of C++ is circa 2000 and there were still a lot of very bad compilers at the time so some of my coding style is to avoid problems in ancient compilers.  Things like letting compiler do default things rather than implement the in source file resulted in blotted binaries with mulple vtable copies and other horrible.  

Looking over most looks reasonable.  Though the curl brace initializers look really odd to me.  I am guessing the weird path with exections was so that I could add source line numbers to the exception which was removed when I added the stack trace linkage.

**Comment by marscher** [comment]
Thanks for taking a look guys!
The question is, if we want to support ancient compilers in an age where c++20(20!) is about to be released. I have to be honest, and admit, that I have to update my c++ knowledge as well. The curly brace initializer just invoke the default constructor, e.g. a raw-pointer (which we actually should stop using) it'd be initialized with a nullptr.

I'm also not favouring an automatic code formatting (right now), as it would break (in a merge-things-back manner) development processes.

**Comment by Thrameos** [comment]
So is this ready for a review at this point or is it still a work in progress?

File: native/python/jp_pythontypes.cpp [hunk 1]
```diff
@@ -358,11 +358,11 @@ JPPyObjectVector::JPPyObjectVector(PyObject* inst, PyObject* sequence)
 {
 	m_Instance = JPPyObject::use(inst);
 	m_Sequence = JPPyObject::use(sequence);
-	size_t n = 0;
+	Py_ssize_t n = 0;
 	if (sequence != nullptr)
 		n = PySequence_Size(m_Sequence.get());
 	m_Contents.resize(n + 1);
-	for (size_t i = 0; i < n; ++i)
+	for (auto i = 0; i < n; ++i)
```
Comment (Thrameos) [inline native/python/jp_pythontypes.cpp:365]: This is a case where I really don't like the use of auto.   Read literally it says ``auto i = 0;``  but ``0`` has no inherent type info.   I am guessing the compiler  is supposed to guess from the comparison statement.   

I think in this case I would prefer being explicit.   It is fine to use auto when we are doing ``auto v = (Type) u;`` where the type of the variable is clear, but letting it simply guess on context from another statement seems unnecessary.

**Comment by Thrameos** [comment]
My general opinion on new compilers like C++20 when supporting a library which is supposed to work from Python 3.7 to current, is that we should stick with features that were considered to be well established when the oldest binary was released.   "Well established" is of course hard to define.   When I was doing gtkmm in late 1990s, I would have to find the common features from Sun, HPUX, gcc,  visual studio and a hand full of others.   There were plenty of useful features in C++ that would make life easier, but inevitably one of the compilers would have a bug of some kind (and that meant finding one of those machines in a lab somewhere and trying to debug the problem there).   Thus I ended deliberately avoiding anything that was near the edge (even when it made for cruddy looking code).   

I still think that we should keep with the same philosophy, though as I don't do as heavy C++ development, I will have to defer to others as to what is well established and what is marginal.   While doing the threading tests, I used C11 features only to discover they don't exist on Macos and that was stuff adopted over 10 years ago.   

So long story short, don't go hog wild using every new language feature.

File: native/python/jp_pythontypes.cpp (hunk 1 above)
Comment (marscher) [reply native/python/jp_pythontypes.cpp]: I used decltype(n) to let the compiler use the type of the condition variable.

**Comment by Thrameos** [comment]
I see codecov is mad about a bunch of missing test cases.  I guess I will take another shot at improving coverage after this one goes in.

File: native/common/include/jp_array.h [hunk 2]
```diff
@@ -23,21 +23,21 @@ class JPArray;
...
-	void *m_Memory;
-	Py_buffer m_Buffer;
+	void *m_Memory{};
+	Py_buffer m_Buffer{};
 	int m_RefCount;
-	Py_ssize_t m_Shape[5];
-	Py_ssize_t m_Strides[5];
-	jboolean m_IsCopy;
-	jboolean m_Owned;
+	Py_ssize_t m_Shape[50]{};
```
Comment (marscher) [inline native/common/include/jp_array.h:37]: I think I made a mistake here, the buffer increased 10 fold.

Comment (marscher) [inline native/common/include/jp_array.h:37]:
```suggestion
	Py_ssize_t m_Shape[5]{};
```

**Comment by Thrameos** [comment]
I am hoping to get out 1.5 this month.  Is this PR in good shape?

**Comment by Thrameos** [comment]
@marscher should I try to get this included in 1.5

**Comment by marscher** [comment]
Sorry for the delay. I guess this is in a good shape - it has been a while since I looked into it. One major change is that JPypeExceptions are now children of std::runtime_error. Honestly I cannot tell if this has any unwanted side effects.
